src/components/generator/generatorFunction.js

```diff
@@ -1,131 +1,147 @@
-      sumOfDifficulties += randomTricks[i].difficultyLevel;
-    }
-    else {
-      // take an random trick
-      randomTricks[i] = possibleTricks[Math.floor(Math.random()*possibleTricks.length)];
-    }
+  return true
+}
+
+/**
+ * Find the first trick depending on whether it needs to be from a certain starting position or not.
+ */
+function firstTrick(availableTricks, isStartPositionFixed, startPosition) {
+  if (!isStartPositionFixed) {
+    return availableTricks[randomIndex(availableTricks.length)]
+  }
+  let possibleTricks = availableTricks.filter(trick => trick.startPos === startPosition)
+  if (possibleTricks.length === 0) {
+    throw "The selected options do not contain any trick with the specified starting position."
   }
+  return possibleTricks[randomIndex(possibleTricks.length)]
+}
 
-  //check integrety of combo
-  for (let i = 1; i < randomTricks.length; i++) {
-    let prev = randomTricks[i - 1];
-    let trick = randomTricks[i];
-    if (prev.endPos !== trick.startPos && !arePositionsSimilar(trick.startPos, prev.endPos) && !allowTransitions) {
-      return alert("trick generator is broken");
-    }
+/**
+ * This adds a heuristics which chooses tricks depending on how it fits towards the expected avgDifficulty,
+ * but the difficultyLevels are modified with a random value of a Normal Distribution (which is depending on how
+ * extreme the avgDifficulty is), to have some additional randomness and not get always the same combo.
+ * Original Author: weberax
+ */
+function randomTrickOfAvgDifficulty(possibleTricks, maxDifficulty, avgDifficulty, sumOfDifficulties, i) {
+  for (let j = 0; j < possibleTricks.length; j++) {
+    let min = -maxDifficulty + 2 * Math.abs(maxDifficulty / 2 - avgDifficulty)
+    let max = maxDifficulty - 2 * Math.abs(maxDifficulty / 2 - avgDifficulty)
+    possibleTricks[j].randomDiff = possibleTricks[j].difficultyLevel + randomNormalDist(min, max);
   }
+  possibleTricks.sort(
+      (a, b) => {
+        let difficultyScoreA = Math.abs(avgDifficulty - (sumOfDifficulties + a.randomDiff) / (i + 1));
+        let difficultyScoreB = Math.abs(avgDifficulty - (sumOfDifficulties + b.randomDiff) / (i + 1))
+        return difficultyScoreB - difficultyScoreA;
+      }
+  )
+  return possibleTricks[possibleTricks.length - 1]
+}
+
+/**
+ * Random value with a Normal Distribution.
+ * Calculated through a Box-Muller transformation, based on this thread:
+ * https://stackoverflow.com/questions/25582882/javascript-math-random-normal-distribution-gaussian-bell-curve
+ */
+function randomNormalDist(min, max) {
+  let u = 1 - Math.random()
+  let v = 1 - Math.random()
+  let rand = Math.sqrt( -2.0 * Math.log( u ) ) * Math.cos( 2.0 * Math.PI * v )
 
-  return randomTricks;
+  rand = rand / 10.0 + 0.5
+  rand = Math.max(0, Math.min(rand, 1))
+
+  return (max - min) * rand + min
 }
+
+/**
+ * Get a random index for a list of the given length. Index is in [0, length-1]
+ */
+function randomIndex(length) {
```

Comment:
Looking at the usages, it might make sense to accept an Array and return the value directly.

e.g. 
```js
function getRandomEntryFromArray(array) {
   return  array[Math.floor(array.length * Math.random())]
}

Reply:
Accepted